module13and14/LoopCanvas/LoopCanvas/Models/Block.swift

```diff
@@ -0,0 +1,179 @@
+//
+//  BlockModel.swift
+//  LoopCanvas
+//
+//  Created by Peter Rice on 6/8/24.
+//
+
+import Foundation
+import SwiftUI
+
+class Block: ObservableObject, Identifiable, Codable {
+  // Persistent props
+  @Published var id: Int
+  @Published var location: CGPoint
+  @Published var color: Color
+
+  var blockGroupGridPosX: Int?
+  var blockGroupGridPosY: Int?
+  var loopURL: URL?
+  let icon: String
+  var relativePath: String?
+
+  // Transient state props
+  @Published var visible = true
+  @Published var dragging = false
+
+  let normalColor: Color
+  let highlightColor: Color = .yellow
+  let isLibraryBlock: Bool
+
+  weak var blockGroup: BlockGroup?
+  var isPlaying = false
+  var loopPlayer: LoopPlayer?
+
+  static var blockIdCounter: Int = 0
+  static func getNextBlockId() -> Int {
+    let blockId = blockIdCounter
+    blockIdCounter += 1
+    return blockId
+  }
+
+  init(
+    id: Int,
+    location: CGPoint,
+    color: Color,
+    icon: String,
+    visible: Bool = false,
+    loopURL: URL? = nil,
+    relativePath: String? = nil,
+    isLibraryBlock: Bool = false
+  ) {
+    self.id = id
+    self.location = location
+    self.color = color
+    self.icon = icon
+    self.relativePath = relativePath
+    self.normalColor = color
+    self.visible = visible
+    self.loopURL = loopURL
+    self.isLibraryBlock = isLibraryBlock
+  }
+
+  func tick(step16: Int) {
+    if step16 % 4 == 0 && isPlaying {
+      color = highlightColor
+    } else if color != normalColor {
+      color = normalColor
+    }
+  }
+
+  // Codable implementation
+
+  // TODO - add an enum for the icons we support as well
+  enum Colors: String, Codable, CaseIterable {
```

Comment:
This was all confusing to me and I can't see what it adds over just using `Color` structures.

Reply:
So I think what I was trying to do here is to make constant colors Codable to JSON by their constant color name (and I guess have flexibility to define new ones, tho not used). So have JSON output { 'color': 'orange' } etc.. And also sort of model a pallet of possible block colors as an enum (was going to do the same for the symbol icons).